### PR TITLE
optional chainId (extracted from #4932)

### PR DESCRIPTION
**Pull request by hackmod**
from #4932
## Description

This PR separates all other stuff except optional chainid support.

## Done
- [x] fixed custom Rpc form to support optional chainid.
  - [x] both old-ui and new-ui
- [x] fixed custom Rpc css style
- [x] show optional ticker symbol + nickname
- [x] hide/show all optional parameters (chainId, symbol and nickname)
  - [x]  only new-ui supported

## See also
- https://github.com/MetaMask/metamask-extension/pull/4516#issuecomment-406096289
- PR #5130 by @chrisfranko

## Screen shot
image

### CODE REVIEW

**Review by ghost — DISMISSED** [review]
Much better, thanks 👍

**Comment by hackmod** [comment]
rebased and fixed defaultValue bug.

**Comment by hackmod** [comment]
@buhrmi any suggestions are welcome 😃

**Comment by buhrmi** [comment]
@hackmod well first of all the failing test needs to be fixed :p

**Comment by buhrmi** [comment]
I've just tested it. I was able to deploy and interact with a contract on the classic network within remix. Seems to work fine and good to go :)

**Comment by chrisfranko** [comment]
Launched a contract using reminx and the public https://node.expanse.tech node with chain_id 2 perfectly fine.

**Review by ghost — DISMISSED** [review]
When merge?

**Comment by bdresser** [comment]
Hi @hackmod @eosclassicteam, thanks again for your patience with this. We appreciate your work and want to make sure it lands with the appropriate implementation that makes it useable and minimally disruptive.

After chatting with the team, here's what we think makes sense:

chain-id-new

#### UI
* The default view in the settings menu should remain the same, all additional options should be hidden behind the `Show Advanced Options` link
* Header for this section should update to `New Network`
* Clicking `Show Advanced Options` link should reveal three optional fields 

#### Functionality
* `Symbol` should update the currency symbol from `ETH` to whatever the user provides
* `Nickname` should determine how the network is named in the Networks dropdown (top-right)
* No field for block explorer yet, we'd like to think about how to communicate the template format 
* We would *not* like to change our default pricing APIs from Infura (for ETH) and Balanc3 (for tokens). If the user has specified a custom network, it's acceptable to use cryptocompare, but the PR should not change our default behavior here. 

#### Future improvements
* We'd eventually like to move in a direction where dapps can suggest a new network to a use (so users don't have to fiddle with these settings). If you're interested, we'd appreciate your feedback on this proposal: https://github.com/MetaMask/metamask-extension/issues/5101
* Currently the network drop-down menu remembers the last three networks a user joined and drops older ones. We plan to update this functionality to remember all networks a user has joined and allow them to remove whichever they'd like (similar to how imported accounts work now) https://github.com/MetaMask/metamask-extension/issues/5246

Does this spec make sense? Sorry we haven't communicated these needs clearly until now, and thanks for bearing with us as we work through this together. I'll do my best to keep this top of mind for the team until we have something merged. And thank you for all your work! (cc @danfinlay)

**Comment by hackmod** [comment]
@bdresser // thank you for your support :)
I'm already prepared with extra fixes like as ticker symbol/price api, will follow your suggestion asap!!

**Comment by hackmod** [comment]
rebased, and follow additional suggestion/recent changes
- [x] optional symbol / optional nickname added
- [x] use cryptocompare api conditionally
- [ ] support deposit with ShapeShift
- [ ] fix remain bugs
## screen shots
image

image

**Comment by bdresser** [comment]
@hackmod great, thank you! is this ready for review, or are you still fixing some bugs?

**Comment by hackmod** [comment]
@bdresser // ready to review now if no further suggestions exist :)

- it looks, almost all price conversion function looks already *almost* prepared. I've fixed a few things to support custom symbol/ticker price, except token related.
- deposit with ShapeShift feature is not incorporated.

my next steps in my mind after merging this PR is as follows.
- supporting network logos to distinguish with the current selected network more clear, like as https://github.com/kvhnuke/etherwallet/pull/2009 PR
- and fix for hardware wallets

**Review by ghost — CHANGES_REQUESTED** [review]
@hackmod  @bdresser Functionality works fine however there is no verified price source for cryptocompare api ( Can't get an info for which exchange they use for their API ), for example I am getting wrong price information for our coin symbol although it is not trading anywhere https://min-api.cryptocompare.com/data/price?fsym=EOSC&tsyms=BTC , using the API should be avoided in order to prevent confusion between our users. I recommend using Coinmarketcap for transparency

**Comment by eosclassicteam** [comment]
@hackmod Here is the api wrapper for CMC in js https://github.com/tiaanduplessis/coinmarketcap-api you can use this instead

**Comment by hackmod** [comment]
@eosclassicteam // i think that is not an issue on this PR.
you'd better request cryptocompare.com about this issue. cryptocompare.com support almost all exchanges except some minor exchanges.
(and there is a paid service which grants higher API rate limits on it, but I think cryptocompare.com is good enough for all ethereum clones.)

**Comment by bdresser** [comment]
If cryptocompare isn't reliable, then I'm in favor of using a more mainstream and consistent pricing API as the backup to balanc3. 

We're trying to squash some bugs on our latest release, so please excuse the delay (again), hope to have this reviewed soon. Thanks for your patience.

**Review by estebanmino — COMMENTED** [review]
Great PR thanks for this @hackmod. I started to review and test it adding a local RPC and I had two bugs.
The first one is logging in with custom RPC as network.
buglogin
The second I believe that is related to the state, maybe something missing there. This is not only happening on this action, every part when a view is rendered.
bugchangingticusti omroc
I tried both cases with a stable version of MetaMask and are working as expected, so it's related to these changes I believe.

File: ui/app/components/pages/settings/settings.js [hunk 1]
```diff
@@ -187,39 +191,104 @@ class Settings extends Component {
...
+            ]),
           ]),
         ]),
       ])
     )
   }
 
-  validateRpc (newRpc) {
+  showAdvancedOptionsLink () {
+    return h('span.settings__advanced-link', {
```
[the remark this answers is not in the document]
Comment (hackmod) [reply ui/app/components/pages/settings/settings.js]: thank you for your review.

there are some changes on repo. I have to rebase/merge and follow with recent changes asap.
Thanks again!!

**Comment by whymarrh** [comment]
@hackmod in addition to @estebanmino's review above there are a few conflicts that need resolving so we can continue reviewing this (a rebase on `develop`)

File: ui/app/components/pages/settings/settings.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hackmod) [reply ui/app/components/pages/settings/settings.js]: fixed `conversionRate(undefined)` case and rebased again

File: ui/app/components/send/currency-display/tests/currency-display.test.js [hunk 2]
```diff
@@ -37,54 +51,58 @@ describe('', () => {
...
         onChange={handleChangeSpy}
-      />, {context: {t: str => str + '_t'}})
+      />, createContext())
 
     const input = wrapper.find('input')
     input.simulate('focus')
     input.simulate('change', { target: { value: '100' } })
 
-    assert.equal(wrapper.state().valueToRender, '100')
+    // assert.equal(wrapper.state().valueToRender, '100') // FIXME
```
Comment (hackmod) [inline ui/app/components/send/currency-display/tests/currency-display.test.js]: I'm not sure why it does not work as expected. anyway, simply ignore for now.

Comment (eosclassicteam) [reply ui/app/components/send/currency-display/tests/currency-display.test.js]: @hackmod LOL, is that related with price conversion? 😂

**Comment by eosclassicteam** [comment]
My personal opinion for custom rpc is, we don't need to show fiat value for them. Right now we can only focus "optional chain id" support only, and we can work for price api later.

File: ui/app/components/send/currency-display/tests/currency-display.test.js (hunk 2 above)
Comment (hackmod) [reply ui/app/components/send/currency-display/tests/currency-display.test.js]: simply fixed my mistake

**Comment by hackmod** [comment]
## screenshot
conversion not supported case (Callisto) - (it is just cryptocompare's fault.)
image

conversion supported case (ETC)
image

image

**Comment by bdresser** [comment]
looks like @hackmod fixed a couple issues @estebanmino or @whymarrh if you have another chance to take a look!

**Comment by alextsg** [comment]
Hi @hackmod , we're currently working on closing out issue #4510 (applying the currency toggle throughout the entire application, and not just the Confirm screen) by the end of the week, which will help de-emphasize fiat values and will likely interact with your changes. We'll review this again once that's completed.

**Review by jason-xie-123 — DISMISSED** [review]

**Comment by alextsg** [comment]
@hackmod There are now several new components used for displaying and inputting currency values. Hopefully they're easy enough to use/modify, but I can provide some help with them if needed.

Can you also provide a description/documentation of the new `fromCurrency` property that's being added? The naming seems a little misleading because it's implying some kind of conversion happening, so it's confusing why that's something that should be kept in state.

**Comment by hackmod** [comment]
> Can you also provide a description/documentation of the new `fromCurrency` property that's being added? The naming seems a little misleading because it's implying some kind of conversion happening, so it's confusing why that's something that should be kept in state.

`fromCurrency` is already used at https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/conversion-util.js#L9

but, as you can see, `fromCurrency` was selected without a deep insight into the current metamask source code.

1. in my PR, `fromCurreny` is used to get the conversion ratio of the selected ethereum fork chains.
2. and it is used as a display property to replace the 'ETH' symbol for customRpc cases.

I'm not much familiar with react programming, any suggestions are welcomed!

anyway, thank you for your hard work! I will follow up and rebase asap!

**Comment by hackmod** [comment]
image

image

**Comment by whymarrh** [comment]
@hackmod I've added a few commits to this PR that we think help the implementation be a bit more generic. I think the commits speak for themselves but let me know if you they don't make sense. The gist of the changes are 1) `fromCurrency` in the `CurrencyController` is better named `nativeCurrency` (47c4508) and 2) the changes in the `CurrencyDisplay` container can be avoided if we pass in the chain's native currency when needed (99acde4).

As-is, this looks good changes-wise. @estebanmino had pointed out some issues with the login above that I think still persist with these changes (maybe with the RPC URL you shared in the OP). Do you have a suggested set of RPC information we can use to do a proper QA pass? I'm seeing a lot of connection issues when using information you've shared so far.

**Comment by estebanmino** [comment]
about @whymarrh comments, I tried it out and I don't have found any issue as the one I mentioned before, for me seems that is working both with local ganache-cli and the information provided, when/where are you having those issues?. I agree that it would be very useful to have more RPC information.

File: ui/app/components/transaction-activity-log/transaction-activity-log.component.js [hunk 3]
```diff
@@ -41,16 +41,17 @@ export default class TransactionActivityLog extends PureComponent {
...
 
   renderActivity (activity, index) {
-    const { conversionRate } = this.props
+    const { conversionRate, nativeCurrency } = this.props
     const { eventKey, value, timestamp } = activity
     const ethValue = index === 0
       ? `${getValueFromWeiHex({
         value,
-        toCurrency: ETH,
+        nativeCurrency,
```
Comment (whymarrh) [inline ui/app/components/transaction-activity-log/transaction-activity-log.component.js]: Whoops, this should be `fromCurrency: nativeCurrency`

File: ui/app/components/transaction-activity-log/transaction-activity-log.component.js [hunk 4]
```diff
@@ -41,16 +41,17 @@ export default class TransactionActivityLog extends PureComponent {
...
+        toCurrency: nativeCurrency,
         conversionRate,
         numberOfDecimals: 6,
-      })} ${ETH}`
-      : getEthConversionFromWeiHex({ value, toCurrency: ETH, conversionRate })
+      })} ${nativeCurrency}`
+      : getEthConversionFromWeiHex({ value, nativeCurrency, toCurrency: nativeCurrency, conversionRate })
```
Comment (whymarrh) [inline ui/app/components/transaction-activity-log/transaction-activity-log.component.js]: Whoops, this should be `fromCurrency: nativeCurrency`

**Review by whymarrh — CHANGES_REQUESTED** [review]

**Comment by whymarrh** [comment]
> when/where are you having those issues?

They're intermittent so I believe they're related to the network connectivity of the added RPC URL. I'm cool with the state of this PR as that's up to what the user adds as their desired network.

**Review by whymarrh — APPROVED** [review]

**Review by alextsg — APPROVED** [review]

**Comment by pyskell** [comment]
> > when/where are you having those issues?
> 
> They're intermittent so I believe they're related to the network connectivity of the added RPC URL. I'm cool with the state of this PR as that's up to what the user adds as their desired network.

@whymarrh I run the ethereumclassic.network RPC, what issues are you seeing? Intermittent connectivity problems or any specific errors?

**Comment by filips123** [comment]
How to rename custom network **after** creation?

**Comment by whymarrh** [comment]
@pyskell I can't remember (sorry!) but I do use that for testing and I'll be sure to mention anything that I run into

@filips123 remove and re-add is the only option right now

**Pull request merged**